### PR TITLE
Skiko native experimental

### PR DESCRIPTION
**Pull request by alexander-gorshenev**

### CODE REVIEW

File: skiko/buildSrc/src/main/kotlin/properties.kt [hunk 1]
```diff
@@ -135,4 +135,8 @@ object SkikoArtifacts {
     val commonArtifactId = "skiko-jvm"
     fun runtimeArtifactIdFor(os: OS, arch: Arch) =
         "skiko-jvm-runtime-${targetId(os, arch)}"
+    fun nativeRuntimeArtifactIdFor(os: OS, arch: Arch) =
+        "skiko-native-runtime-${targetId(os, arch)}"
+    fun nativeSkiaInteropArtifactIdFor(os: OS, arch: Arch) =
```
Comment (olonho) [inline skiko/buildSrc/src/main/kotlin/properties.kt:140]: Why do we need second artifact? `skiko-native-runtime` shall be sufficient.

File: skiko/gradle.properties [hunk 2]
```diff
@@ -1,5 +1,6 @@
 kotlin.code.style=official
 deploy.version=0.0.0
+org.jetbrains.kotlin.native.version=1.5.20-dev-5723
```
Comment (olonho) [inline skiko/gradle.properties:3]: `Project property 'org.jetbrains.kotlin.native.version' is deprecated. Please use 'kotlin.native.version' instead.`

File: samples/SkiaNativeSample/settings.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+pluginManagement {
```
Comment (olonho) [inline samples/SkiaNativeSample/settings.gradle.kts:1]: Gradle wrappers are missing in PR.

File: skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/PlatformOperations.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    fun isFullscreen(component: Component): Boolean
+    fun setFullscreen(component: Component, value: Boolean)
+    fun getDpiScale(component: Component): Float
+     */
+    fun createRedrawer(layer: HardwareLayer, properties: SkiaLayerProperties): Redrawer
+}
+
+internal val platformOperations: PlatformOperations by lazy {
+    //when (hostOs) {
+        /*OS.MacOS -> */object: PlatformOperations {
```
Comment (olonho) [inline skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/PlatformOperations.kt:18]: Let's discuss how to do that better.

File: skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/redrawer/MacOsOpenGLRedrawer.kt [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+import platform.CoreVideo.CVTimeStamp
+import platform.OpenGLCommon.CGLContextObj
+import platform.OpenGLCommon.CGLPixelFormatObj
+import platform.OpenGLCommon.CGLSetCurrentContext
+import platform.QuartzCore.CAOpenGLLayer
+import platform.QuartzCore.*
+import platform.darwin.dispatch_async
+import platform.darwin.dispatch_get_main_queue
+import kotlin.system.getTimeNanos
+
```
Comment (olonho) [inline skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/redrawer/MacOsOpenGLRedrawer.kt:18]: Please clean it up, cyrrent AWT remnants make it look awful.

File: skiko/src/nativeInterop/cinterop/skia.def [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    -DSK_SUPPORT_OPENCL=0 \
+    -DSK_UNICODE_AVAILABLE
+
+headers = include/core/SkCanvas.h include/core/SkData.h include/core/SkImage.h include/core/SkMath.h include/core/SkStream.h include/core/SkSurface.h include/core/SkSurfaceProps.h include/core/SkTypes.h include/core/SkPaint.h include/core/SkPictureRecorder.h include/core/SkImageInfo.h include/core/SkColor.h include/core/SkColorSpace.h include/gpu/GrDirectContext.h include/gpu/GrBackendSurface.h include/gpu/GrConfig.h include/gpu/gl/GrGLInterface.h
+headerFilter =  include/core/SkCanvas.h include/core/SkData.h include/core/SkImage.h include/core/SkMath.h include/core/SkStream.h include/core/SkSurface.h include/core/SkSurfaceProps.h include/core/SkTypes.h include/core/SkPaint.h include/core/SkPictureRecorder.h include/core/SkImageInfo.h include/core/SkColor.h include/core/SkColorSpace.h include/gpu/GrDirectContext.h include/gpu/GrBackendSurface.h include/gpu/GrConfig.h include/gpu/gl/GrGLInterface.h
+
+# const overloads are not supported yet
+excludedFunctions = saveLayer mapRect ptr priv
+
+linkerOpts.osx = -framework CoreServices -framework CoreText -framework CoreGraphics -framework Metal
```
Comment (olonho) [inline skiko/src/nativeInterop/cinterop/skia.def:25]: Previous code gives impression that we use OpenGL for rendering, while this code links Metal. What is actually correct?

Comment (alexander-gorshenev) [reply skiko/src/nativeInterop/cinterop/skia.def:25]: OpenGL is added by K/N automatically. 
While Metal is still needed for libskia.a to resolve symbols.
```
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_MTLRenderPipelineColorAttachmentDescriptor", referenced from:
      objc-class-ref in libskia.a(gpu.GrMtlPipelineStateBuilder.o)
  "_OBJC_CLASS_$_MTLRenderPipelineDescriptor", referenced from:
      objc-class-ref in libskia.a(gpu.GrMtlPipelineStateBuilder.o)
  "_OBJC_CLASS_$_MTLVertexDescriptor", referenced from:
      objc-class-ref in libskia.a(gpu.GrMtlPipelineStateBuilder.o)
  "_OBJC_CLASS_$_MTLTextureDescriptor", referenced from:
      objc-class-ref in libskia.a(gpu.GrMtlUtil.o)
      objc-class-ref in libskia.a(gpu.GrMtlGpu.o)
      objc-class-ref in libskia.a(gpu.GrMtlAttachment.o)
      objc-class-ref in libskia.a(gpu.GrMtlRenderTarget.o)
      objc-class-ref in libskia.a(gpu.GrMtlTextureRenderTarget.o)
  "_OBJC_CLASS_$_MTLRenderPassDescriptor", referenced from:
      objc-class-ref in libskia.a(gpu.GrMtlGpu.o)
      objc-class-ref in libskia.a(gpu.GrMtlOpsRenderPass.o)
  "_OBJC_CLASS_$_MTLDepthStencilDescriptor", referenced from:
      objc-class-ref in libskia.a(gpu.GrMtlDepthStencil.o)
  "_OBJC_CLASS_$_MTLStencilDescriptor", referenced from:
      objc-class-ref in libskia.a(gpu.GrMtlDepthStencil.o)
  "_OBJC_CLASS_$_MTLSamplerDescriptor", referenced from:
      objc-class-ref in libskia.a(gpu.GrMtlSampler.o)
ld: symbol(s) not found for architecture x86_64
```

File: skiko/gradle.properties (hunk 2 above)
Comment (alexander-gorshenev) [reply skiko/gradle.properties]: chenged

File: skiko/buildSrc/src/main/kotlin/properties.kt (hunk 1 above)
Comment (alexander-gorshenev) [reply skiko/buildSrc/src/main/kotlin/properties.kt:140]: Discussed offline. We have to have two artifacts for now.

File: samples/SkiaNativeSample/settings.gradle.kts (hunk 3 above)
Comment (alexander-gorshenev) [reply samples/SkiaNativeSample/settings.gradle.kts:1]: added

File: skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/redrawer/MacOsOpenGLRedrawer.kt (hunk 5 above)
Comment (alexander-gorshenev) [reply skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/redrawer/MacOsOpenGLRedrawer.kt:15]: Removed many porting leftovers

File: skiko/build.gradle.kts [hunk 7]
```diff
@@ -154,6 +156,33 @@ kotlin {
...
     }
 
+    val nativeTarget = when (target) {
+        "macos-x64" -> macosX64()
+        // "linux-x64" -> linuxX64()
+        // "windows-x64" -> mingwX64()
+        else -> null
+    }
+    nativeTarget?.apply {
+        compilations.getByName("main") {
```
Comment (olonho) [inline skiko/build.gradle.kts:166]: Does native builds by default now? Let's put it under flag for now.

File: skiko/buildSrc/src/main/kotlin/properties.kt (hunk 1 above)
Comment (olonho) [reply skiko/buildSrc/src/main/kotlin/properties.kt:140]: Ack

File: skiko/build.gradle.kts (hunk 7 above)
Comment (olonho) [reply skiko/build.gradle.kts:166]: I.e. `./gradlew -Pskiko.native.enabled=true` and `if (project.properties.get("skiko.native.enabled") == "true")` protecting native tasks.

File: skiko/build.gradle.kts [hunk 8]
```diff
@@ -154,6 +156,33 @@ kotlin {
...
+            val skia by cinterops.creating {
+                defFile("src/nativeInterop/cinterop/skia.def")
+
+                val skiaDir = skiaDir.get().absolutePath
+                val kotlinNativeDataPath = System.getenv("KONAN_DATA_DIR")?.let { File(it) }
+                    ?: File(System.getProperty("user.home")).resolve(".konan")
+
+                compilerOpts(
+                     // TODO: Why do we need `include/c++/v1`?
```
Comment (olonho) [inline skiko/build.gradle.kts:175]: Do we really need it C++ headers now?

File: skiko/build.gradle.kts [hunk 9]
```diff
@@ -154,6 +156,33 @@ kotlin {
...
+                     "-I$kotlinNativeDataPath/dependencies/clang-llvm-apple-8.0.0-darwin-macos//include/c++/v1",
+                     "-I$skiaDir"
+                )
+
+                extraOpts("-staticLibrary", "libskia.a")
+                extraOpts("-libraryPath", "$skiaDir/out/Release-x64")
```
Comment (olonho) [inline skiko/build.gradle.kts:181]: Make it `${buildType.id}-${targetArch.id}`.

File: skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/HardwareLayer.kt [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+abstract class HardwareLayer {
+
+    val nsView = NSView(NSMakeRect(0.0, 0.0, 640.0, 480.0))
+
+    // getDpiScale is expensive operation on some platforms, so we cache it
+    private var _contentScale: Float? = null
+    private var isInit = false
+
+
+    init {
```
Comment (olonho) [inline skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/HardwareLayer.kt:18]: Remove empty init.

File: skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/PlatformOperations.kt [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+internal interface PlatformOperations {
+    fun createRedrawer(layer: HardwareLayer, properties: SkiaLayerProperties): Redrawer
+}
+
+internal val platformOperations: PlatformOperations by lazy {
+        object: PlatformOperations {
+                override fun createRedrawer(layer: HardwareLayer, properties: SkiaLayerProperties) = when(renderApi) {
+                    GraphicsApi.SOFTWARE -> error("No software rendering for native yet")
+                    else -> MacOsOpenGLRedrawer(layer, properties)
```
Comment (olonho) [inline skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/PlatformOperations.kt:15]: Why not make it proper?

File: skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/SkikoProperties.kt [hunk 12]
```diff
@@ -0,0 +1,19 @@
+package org.jetbrains.skiko.native
+
+internal object SkikoProperties {
+
+    val vsyncEnabled: Boolean  = true
+    //val vsyncEnabled: Boolean by property("skiko.vsync.enabled", default = true)
```
Comment (olonho) [inline skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/SkikoProperties.kt:6]: Why commented out code is needed?

File: skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/redrawer/MacOsOpenGLRedrawer.kt [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    private var isDisposed = false
+
+    private val drawLayer = MacosGLLayer(layer, setNeedsDisplayOnBoundsChange = true)
+
+    override fun dispose() { 
+        drawLayer.dispose()
+    }
+
+    override fun syncSize() {
+        // TODO: What do we really do here?
```
Comment (olonho) [inline skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/redrawer/MacOsOpenGLRedrawer.kt:29]: At least ```
layer.nsView.frame.useContents {
   drawLayer.setFrame(origin.x.toInt(), ...)
}
```

File: skiko/src/nativeInterop/cinterop/skia.def (hunk 6 above)
Comment (olonho) [reply skiko/src/nativeInterop/cinterop/skia.def:25]: OK.

File: skiko/build.gradle.kts (hunk 7 above)
Comment (alexander-gorshenev) [reply skiko/build.gradle.kts]: done that

File: skiko/build.gradle.kts (hunk 8 above)
Comment (alexander-gorshenev) [reply skiko/build.gradle.kts]: Yes for now. There will be a corresponding fix to native that fixes that.

File: skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/HardwareLayer.kt (hunk 10 above)
Comment (alexander-gorshenev) [reply skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/HardwareLayer.kt]: removed

File: skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/SkikoProperties.kt [hunk 14]
```diff
@@ -1,9 +1,10 @@
 package org.jetbrains.skiko.native
 
+// TODO: jvm version uses java properties here.
+// there is no properties support available for native.
```
Comment (olonho) [inline skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/SkikoProperties.kt:4]: Maybe using env variables (later)?

Comment (alexander-gorshenev) [reply skiko/src/macosX64Main/kotlin/org/jetbrains/skiko/SkikoProperties.kt:4]: later, yes. Will discuss.

File: skiko/build.gradle.kts [hunk 15]
```diff
@@ -154,6 +156,35 @@ kotlin {
         withJava()
     }
 
+    if (project.properties.get("skiko.native.enabled") == "true") {
```
Comment (olonho) [inline skiko/build.gradle.kts:159]: Maybe just an extension property of Project, i.e.
```
val Project.supportNative: Boolean
   get() = properties.get("skiko.native.enabled") == "true"
```

**Review by olonho — APPROVED** [review]

**Pull request merged**